packages/playwright-core/src/server/recorder/recorderCollection.ts

```diff
@@ -46,29 +53,55 @@ export class RecorderCollection extends EventEmitter {
     this._enabled = enabled;
   }
 
-  addAction(action: ActionInContext) {
+  async willPerformAction(actionInContext: ActionInContext): Promise<CallMetadata | null> {
     if (!this._enabled)
-      return;
-    this.willPerformAction(action);
-    this.didPerformAction(action);
+      return null;
+    const mainFrame = mainFrameForAction(this._pageAliases, actionInContext);
+
+    const { action } = actionInContext;
+    const callMetadata: CallMetadata = {
+      id: `call@${createGuid()}`,
+      apiName: 'frame.' + action.name,
+      objectId: mainFrame.guid,
+      pageId: mainFrame._page.guid,
+      frameId: mainFrame.guid,
+      startTime: monotonicTime(),
+      endTime: 0,
+      type: 'Frame',
+      method: action.name,
+      params: traceParamsForAction(actionInContext),
+      log: [],
+    };
+    await mainFrame.instrumentation.onBeforeCall(mainFrame, callMetadata);
+    this._currentAction = actionInContext;
+    return callMetadata;
   }
 
-  willPerformAction(action: ActionInContext) {
+  async didPerformAction(callMetadata: CallMetadata, actionInContext: ActionInContext, error?: Error) {
     if (!this._enabled)
       return;
-    this._currentAction = action;
-  }
 
-  performedActionFailed(action: ActionInContext) {
-    if (!this._enabled)
-      return;
-    if (this._currentAction === action)
+    if (error) {
+      // Do not clear current action on delayed error.
+      if (this._currentAction === actionInContext)
+        this._currentAction = null;
+    } else {
       this._currentAction = null;
+      this._actions.push(actionInContext);
+    }
+
+    this._lastAction = actionInContext;
```

Comment:
Do we want to set `_lastAction` in the error case?

Reply:
Hm, I'll think about what this means. We definitely don't want to glue to the previous one, so maybe we want to reset it.